src/tests/ftest/network/zero_config.py

```diff
@@ -169,32 +201,36 @@ def test_env_set_unset(self):
         :avocado: tags=all,daily_regression,hw,small,zero_config,env_set
         """
         env_state = self.params.get("env_state", '/run/zero_config/*')
-        dev_info = {"ib0": 0, "ib1": 1}
-        exp_iface = random.choice(list(dev_info.keys()))
+
+        # Get the available interfaces and their domains
+        self.get_device_info()
+        exp_iface = random.choice(list(self.dev_info.keys()))
 
         # Configure the daos server
-        self.add_server_manager()
-        self.configure_manager(
-            "server",
-            self.server_managers[0],
-            self.hostlist_servers,
-            self.hostfile_servers_slots,
-            self.access_points)
+        self.setup_servers()
         self.assertTrue(
             self.server_managers[0].set_config_value(
                 "fabric_iface", exp_iface),
             "Error updating daos_server 'fabric_iface' config opt")
         self.assertTrue(
             self.server_managers[0].set_config_value(
-                "pinned_numa_node", dev_info[exp_iface]),
+                "pinned_numa_node", self.dev_info[exp_iface]["numa"]),
```

Comment:
So, if I understand the code correctly, you're configuring the servers to use a particular interface based on its numa node, and then expecting to see traffic on the same interface on the clients, right?

If that's correct, then I'm not sure that this is a particularly resilient approach. It may work on this particular cluster/configuration, but as far as I'm aware there's no forced correlation between the server socket and the client socket. In fact, the client interface is likely to change based on the numa affinity of the client process.

The way the client zero-config works is that the server sets a "hint" for the client based on what it knows about its own network config: https://github.com/daos-stack/daos/blob/master/src/control/server/server.go#L306 Note that it doesn't specify an address directly; it just says "here's the type of interface (ethernet/ib/etc) I'm using, and here's the provider I'm using".

It's then up to the agent to select a client-side interface for the application based on its numa affinity and the server's hint for device class: https://github.com/daos-stack/daos/blob/master/src/control/cmd/daos_agent/fabric.go#L120 ... So the first network interface that matches the server's interface class and the client application's numa affinity will be selected. Furthermore, if there is more than one interface that could match, then subsequent selections will be round-robined among the matching interfaces.

What might be a better approach is to check all of the interfaces that could be selected for the client and verify that at least one of them sent and received some traffic. That still feels a bit shaky though... Would it be reasonable to set expected values for read/write throughput that are in the same ballpark as the actual values, based on the daos_racer params?

Reply:
Before we get into the accuracy/correctness/robustness of this test, this particular PR is just fixing the existing test so that it will work on CI clusters that have Mellanox adapters.

I agree the test could be better, but it probably should addressed in a separate ticket/PR.  It seems that besides starting the daos server with a random ib* interface, the test is really just verifying that setting/not setting the OFI_INTERFACE environment variable has no impact on being able to run daos_racer.

Reply:
Fair enough re: accuracy, but my point is just that you may get spurious failures if the agent-selected interface on the client doesn't match the interface that the test is expecting to be used.

Reply:
Yes, definitely a possibility.  If there is a way to query the agent for the interface it has selected does it make sense for the test to do this or should this test be verifying that the agent has chosen only one of the possible valid interfaces?

Reply:
My instinct would be to do as suggested, collect all valid interfaces for the provider on the client node and ensure we used one of them. Seems like the simplest solution.

I also thought of having the test do something with the manual fabric setup in the agent config to ensure the agent sees only one interface, but that seems more complicated.

Reply:
Sorry, I did not catch the point. As my understand, daos_racer is totally application level, nothing related with how the lower network interface will be chosen. What will happen if we enumerate all valid interfaces and setup/use them one by one (with OFI_INTERFACE if applicable)?